### PR TITLE
Add a 2x2 grid layout function shortcut

### PR DESCRIPTION
**Pull request by mirubiri**
@sdwolfz Here you got the window shortcut for the 2x2 grid ;)

### CODE REVIEW

**Comment by sdwolfz** [comment]
Awesome! I'll take a look and merge it this evening.

File: layers/+spacemacs/spacemacs-defaults/funcs.el [hunk 1]
```diff
@@ -559,6 +559,17 @@ in a split window to the right."
   (split-window-horizontally)
   (other-window 1))
 
+(defun spacemacs/layout-grid ()
```
Comment (mirubiri) [inline layers/+spacemacs/spacemacs-defaults/funcs.el]: I'm not sure if the name of the function should include a `2x2` or this is ok.

Comment (sdwolfz) [reply layers/+spacemacs/spacemacs-defaults/funcs.el]: It's OK as is, I don't realy think other grid layouts might be practical due to the small size of windows.

Sometimes I code with a 2x2 layout with an extra neotree open on the side and it becomes hard to read if I have to do many changes in just one file. But it's great for small one/two liners across multiple files.

Comment (mirubiri) [reply layers/+spacemacs/spacemacs-defaults/funcs.el]: Yup ok, that's my way of thinking too :)

**Comment by bmag** [comment]
Hey, I saw this nice addition and I've got a small comment. It's meant only to share some knowledge, not as a request for changes (since @sdwolfz already said he'd look and merge it):

FWIW, nearly all window-splitting functions return the new window as their return value, and `split-window` can split any window (not just the currently selected window), so you could use it to write code that doesn't rely on `windmove` or on changing the selected window.

Just as a taste, this code splits the current window into a 2x2 windows as well:
```elisp
(split-window (split-window-right) nil 'below)  ; split selected window to the right, then split that window below
(split-window-below)
```

**Comment by sdwolfz** [comment]
@bmag That's some really nice elisp!

I was also thinking about implementing a variable that controls whether the filetree (neotree/treemacs) window is deleted or not. Anyway I can't wait to get home and play around with this.

**Comment by mirubiri** [comment]
@bmag Nice! I don't mind to change the code it that works!

**Comment by bmag** [comment]
> a variable that controls whether the filetree (neotree/treemacs) window is deleted or not

Maybe utilize the prefix argument for that? So `M-x layout-grid` doesn't delete neotree/treemacs but `C-u M-x layout-grid` deletes also neotree/treemacs.
(in Spacemacs it's `SPC SPC` and `SPC u` instead of `M-x` and `C-u`, I used the latter for the sake of being clear)

Also, you probably know that normally neotree/treemacs prevent their window from being deleted by `delete-other-windows`. However, If they are well behaved, then
```elisp
(let ((ignore-window-parameters t))
  (delete-other-windows))
```
will override the protections and delete also the neotree/treemacs window.

**Comment by bmag** [comment]
> currently all the layout functions defined here remove the neotree window.

I haven't used them in a while, maybe I'm out of the loop...

**Comment by fiveNinePlusR** [comment]
> currently all the layout functions defined here remove the neotree window.

They might do that now but should they do that moving forward? I am for leaving the file tree up as the standard behavior as default because that's more of a protected class of windows imo. Same with the imenu list on the right. having a prefix arg close everything would be a good compromise or maybe a even a toggle option?

**Comment by sdwolfz** [comment]
Thanks everyone for the help and insights :heart:!
Keeping the filetree alive seems to be a bit tricky... I'll open up a PR with a possible solution and we can continue the discussion there.

Cherry-picked into develop branch, you can safely delete your branch.

**Pull request closed without merging**